### DISCUSSION TITLE
YOLOv8x: How to Reduce False Positives & Improve Generalization in Table Tennis Ball Detection?

### DISCUSSION THREAD

**Discussion by vaibhav-corp** [ultralytics/ultralytics#21251] [Q&A]
Hi Ultralytics team,

I'm working on a table tennis project and training a YOLOv8x model for ball detection with the goal of eventually building a score counter. However, I’m facing major issues with false positives and generalization.

**Dataset Details:**

1. 5,000 custom annotated images of table tennis balls
2. Includes a variety of lighting conditions, angles, environments, and shot types
3. Images were resized to 640×640 during dataset creation (e.g., Roboflow or similar)
4. Dataset is diverse and covers many typical game scenarios

**Training Configs:**

- Model: YOLOv8x
- Training image size: 1280
- Trained using the Ultralytics standard training pipeline

**Observations:**

1. On one specific type of video (with consistent angle and lighting), the model detects the ball quite well — but still shows false positives (detecting hands, shoes, table edges as balls).
2. On other video types (with different angles, more motion blur, etc.), the model mostly produces false positives only, and often misses the actual ball completely.
3. Interestingly, I had previously trained the same dataset using 640 image size (with YOLOv8x), and while it performed poorly on that one consistent video, it performed better on the other varied videos.
4. This inconsistency makes it hard to build a reliable detection system for scoring. My main goal is to reduce false detections and get stable, generalizable performance across all common table tennis match conditions.

**Questions:**

1. How can I improve generalization across videos with different angles and lighting?
2. How can I reduce false positives, especially for objects like shoes, hands, and net corners?
3. Could video frame-based training or temporal consistency help improve performance?

Any recommendations on whether I should go back to training at 640 vs stick with 1280?

Here’s the [dataset link](https://universe.roboflow.com/vaibhav-rcmge/balls-main-dupli/browse?queryText=&pageSize=50&startingIndex=0&browseQuery=true) in case it’s helpful.

Thanks so much for your work on Ultralytics and YOLO — any tips or suggestions would be really appreciated!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @vaibhav-corp, thank you for your detailed post and for using Ultralytics 🚀! For users tackling custom projects like yours, we recommend reviewing the [Docs](https://docs.ultralytics.com/) for a wide range of [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common training and generalization questions are addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

For custom training and model improvement ❓ discussions (like yours), please include as much information as possible—such as dataset examples, detailed training logs, and anything else you think is relevant. Also, check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), which cover augmentation, overfitting, and ways to boost generalization and reduce false positives.

We invite you to join the Ultralytics community wherever you’re most comfortable. For real-time support, hop onto [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth technical discussions, visit [Discourse](https://community.ultralytics.com/), or share experiences and questions with others on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Make sure you’re running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify you’re not encountering issues already addressed in recent updates:

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of the following verified environments, all with preinstalled dependencies (including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help guide your issue to a resolution. An Ultralytics engineer will review your discussion and assist you soon! 😊

**Reply by glenn-jocher** [reply]
Hi @vaibhav-corp! Great detailed description of your table tennis detection project. Based on your observations, here are some targeted recommendations:

For **reducing false positives** on hands/shoes/table edges, try increasing your `conf` threshold during inference (start with `conf=0.5` or higher) and consider adding hard negative mining - include cropped images of these common false positive objects with no ball annotations in your training data.

For **generalization across different conditions**, your dataset diversity sounds good but consider using stronger augmentation during training with `hsv_h=0.015`, `hsv_s=0.7`, `hsv_v=0.4`, and `degrees=10` to better handle lighting/angle variations. Also try [Test-Time Augmentation](https://docs.ultralytics.com/guides/model-evaluation-insights/#test-time-augmentation-tta) during inference.

Regarding **image size**: Your observation about 640 vs 1280 performance difference suggests the ball size relative to image resolution matters significantly. Since table tennis balls are small objects, try training at 640 but ensure your original images have sufficient resolution before resizing - if balls are too small after resizing, detection becomes much harder.

For **temporal consistency** in videos, consider implementing simple tracking logic post-detection rather than training on video frames - detect on individual frames then apply temporal smoothing or tracking to reduce flickering false positives.

Would you be able to share some example detection results showing the false positives you're seeing? This would help provide more specific guidance.

**Comment by Y-T-G** [comment]
You can increase the confidence threshold to reduce false positives.

> Images were resized to 640×640 during dataset creation (e.g., Roboflow or similar)

This is unnecessary. And it's worse if you're using 640x640 to train with `imgsz=1280` because that just means you're upscaling it to 1280x1280.
